pkg/storage/mvcc_key.go

```diff
@@ -327,3 +327,68 @@ func decodeMVCCTimestampSuffix(encodedTS []byte) (hlc.Timestamp, error) {
 	}
 	return decodeMVCCTimestamp(encodedTS[:encodedLen-1])
 }
+
+// MVCCRangeKey is a versioned key span.
+type MVCCRangeKey struct {
+	StartKey  roachpb.Key
+	EndKey    roachpb.Key
+	Timestamp hlc.Timestamp
+}
+
+// Clone returns a copy of the range key.
+func (k MVCCRangeKey) Clone() MVCCRangeKey {
+	// k is already a copy, but byte slices must be cloned.
+	k.StartKey = k.StartKey.Clone()
+	k.EndKey = k.EndKey.Clone()
+	return k
+}
+
+// Compare returns -1 if this key is less than the given key, 0 if they're
+// equal, or 1 if this is greater. Comparison is by start,timestamp,end, where
+// larger timestamps sort before smaller ones except empty ones which sort first
+// (like elsewhere in MVCC).
+func (k MVCCRangeKey) Compare(o MVCCRangeKey) int {
+	if c := k.StartKey.Compare(o.StartKey); c != 0 {
+		return c
+	}
+	if k.Timestamp.IsEmpty() && !o.Timestamp.IsEmpty() {
+		return -1
+	} else if !k.Timestamp.IsEmpty() && o.Timestamp.IsEmpty() {
+		return 1
+	} else if c := k.Timestamp.Compare(o.Timestamp); c != 0 {
+		return -c // timestamps sort in reverse
+	}
+	return k.EndKey.Compare(o.EndKey)
+}
+
+// String formats the range key.
+func (k MVCCRangeKey) String() string {
+	s := roachpb.Span{Key: k.StartKey, EndKey: k.EndKey}.String()
+	if !k.Timestamp.IsEmpty() {
+		s += fmt.Sprintf("/%s", k.Timestamp)
+	}
+	return s
+}
+
+// Validate returns an error if the range key is invalid.
+func (k MVCCRangeKey) Validate() (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "invalid range key %s", k)
+	}()
+
+	if k.StartKey == nil {
```

Comment:
>nil key is equal to roachpb.KeyMin:

Not when using `==`, so this will allow passing `keys.MinKey`, but I get your point.

Range keys are only valid over MVCC spans. The first span after `KeyMin` is `Meta1Span`, which I think is actually an MVCC span, since range descriptors can have intents. However, all meta1 keys are under `Meta1Prefix`, so I suppose if someone really wants to drop a range tombstone across the meta1 span they can use `Meta1Prefix` as the start key. Maybe we should go the other way and just disallow `len(StartKey) == 0`?

Not sure if this is really a big deal either way though -- the current check at least prevents the field from being accidentally unset, but allows setting it to an empty byte slice if someone really wants to. That seems fine maybe?

Reply:
Looks like `MVCCPut()` rejects writing to an empty key, so I guess we should do the same here

https://github.com/cockroachdb/cockroach/blob/a9e5753535ad5c12ab522580177876a145ac2809/pkg/storage/mvcc.go#L1292-L1294

Reply:
Updated to reject empty start keys.